Adds `remove_if` Function to `linestring_intersection_intermediates`

isVoid opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->

https://github.com/rapidsai/cuspatial/pull/813 adds a struct to hold the intermediate result from `linestring_intersection_with_duplicates`. After discovering the duplicates from the result, we need a way to remove these geometries from the struct, as well as updating the offsets and look-back ids. This PR adds a `remove_if` function to the struct, which models after `thrust::remove_if` that accepts a range of flags, and removes the geometries in the intermediates where the flag is set to 1.

Closes #853 
Depends on #813. Contributes to #765.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

harrism commented:
I reviewed the remove_if function and it looks fine. Could maybe use some more explanatory documentation of the algorithm since it's a fairly complicated function. Will review more fully once dependencies are merged.

isVoid commented:
rerun tests

isVoid commented:
@thomcom I believe I addressed all your reviews regarding `remove_if` you posted in #852. I added more comments per https://github.com/rapidsai/cuspatial/pull/852#discussion_r1048760084. As for https://github.com/rapidsai/cuspatial/pull/852#discussion_r1048718340, I believe the naming is a convention I pulled from thrust and is a common short hand for "remove element if the flag is set". This is similar to `copy_if`.

isVoid commented on cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
+  void remove_if(FlagRange flags, rmm::cuda_stream_view stream)
+  {
+    // The offsets for the geometry array marks the start index for each list.
+    // When geometry are removed, the offset for the next list must
```
```suggestion
    // When geometries are removed, the offset for the next list must
```

isVoid commented on cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
+                            reduced_keys.begin(),
+                            reduced_flags.begin(),
+                            thrust::equal_to<index_t>(),
+                            thrust::plus<index_t>());  // explicitly cast flagss to index_t type
```
```suggestion
                            thrust::plus<index_t>());  // explicitly cast flags to index_t type
```

isVoid commented on cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
+    reduced_keys.resize(thrust::distance(reduced_keys.begin(), keys_end), stream);
+    reduced_flags.resize(thrust::distance(reduced_flags.begin(), flags_end), stream);
+
+    // Use `inclusive_scan` to compute the number of removed geometry in *all* previous lists.
```
```suggestion
    // Use `inclusive_scan` to compute the number of removed geometries in *all* previous lists.
```

isVoid commented on cpp/include/cuspatial/experimental/detail/linestring_intersection_with_duplicates.cuh:
```diff
@@ -166,6 +195,69 @@ struct linestring_intersection_intermediates {
+      intersection_functors::offsets_update_functor{
+        reduced_keys.begin(), reduced_keys.end(), reduced_flags.begin(), reduced_flags.end()});
+
+    // Remove the geometry and corresponding id array per flag.
```
```suggestion
    // Remove the geometries and the corresponding ids per flag.
```

isVoid reviewed (commented):
A couple typos

isVoid commented:
@gpucibot  merge

isVoid commented:
@gpucibot merge
